### PR TITLE
Matrix multiplication (or dot product)

### PR DESCRIPTION
**Pull request by muupan**
This PR adds two functions named ```matmul``` and ```batch_matmul```, both of which do matrix multiplication of two ```Variable```s representing matrices. They also can be used to compute vector inner products.

```matmul``` accepts two ```Variable```s that are 1-D or 2-D arrays. 1-D arrays are treated as column vectors while 2-D arrays are as matrices. Its output is always a 2-D array representing a matrix.

```batch_matmul``` accepts two ```Variable```s that are 2-D or 3-D arrays. 2-D arrays are treated as batches of column vectors while 3-D arrays are as batches of matrices. Its output is always a 3-D array representing a batch of matrices.

Both accept Boolean options ```transa``` and ```transb``` for transposing corresponding argument matrices.

This PR may solve #8.

### CODE REVIEW

File: chainer/functions/matmul.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
+import numpy
+import six
+
+from chainer import cuda
+from chainer import function
+
+
+def _mat_ptrs(a):
+    """Create an array of pointers to matrices
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> Create**s** an array of pointers to matrices**.**

File: chainer/functions/matmul.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    Args:
+        a: batch of matrices in GPU
+    Returns:
+        GPU array of pointers to matrices
+    """
+    return cuda.to_gpu(numpy.arange(
+        a.ptr, a.ptr + a.shape[0] * a.strides[0], a.strides[0],
+        dtype=cuda.cublas.ctypes.c_void_p))
```
Comment (beam2d) [inline chainer/functions/matmul.py]: Import and use `ctypes` directly.

File: chainer/functions/matmul.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+from chainer import cuda
+from chainer import function
+
+
+def _mat_ptrs(a):
+    """Create an array of pointers to matrices
+
+    Args:
+        a: batch of matrices in GPU
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> a: **A** batch of matrices in GPU**.**

File: chainer/functions/matmul.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        a, b, transa, transb = b, a, not transb, not transa
+    a = _as_batch_mat(a)
+    b = _as_batch_mat(b)
+    alpha = numpy.float32(1.0)
+    beta = numpy.float32(0.0)
+    l, m, k = a.shape
+    if transa:
+        m, k = k, m
+    n = b.shape[1] if transb else b.shape[2]
+    with cuda.using_cumisc():
```
Comment (beam2d) [inline chainer/functions/matmul.py]: You do not need to call using_cumisc to just call a function in `cuda.cublas`.

File: chainer/functions/matmul.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        gx0 = _matmul_cpu(
+            gy[0], x[1], transb=not self.transb, transout=self.transa
+            ).reshape(x[0].shape)
+        gx1 = _matmul_cpu(
+            x[0], gy[0], transa=not self.transa, transout=self.transb
+            ).reshape(x[1].shape)
+        return gx0, gx1
+
+    def backward_gpu(self, x, gy):
+        with cuda.using_cumisc():
```
Comment (beam2d) [inline chainer/functions/matmul.py]: The using_cumisc is duplicated with that of _matmul_gpu. Remove either of them.

File: chainer/functions/matmul.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                gy[0], x[1], transb=not self.transb, transout=self.transa
+                ).reshape(x[0].shape)
+            gx1 = _matmul_gpu(
+                x[0], gy[0], transa=not self.transa, transout=self.transb
+                ).reshape(x[1].shape)
+            return gx0, gx1
+
+
+def matmul(a, b, transa=False, transb=False):
+    """Compute matrix multiplication of two arrays.
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> Compute**s** matrix multiplication of two arrays.

File: chainer/functions/matmul.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+    Args:
+        a, b: Variables of 2-D or 1-D arrays.
+            A 2-D array with shape (N, M) is considered as a NxM matrix.
+            A 1-D array with shape (N,) is considered as a Nx1 matrix.
```
Comment (beam2d) [inline chainer/functions/matmul.py]: Do not put two arguments to the same entry. For example:
```
    a (Variable): The left operand of the matrix multiplication.
        A 1-D array of shape (N,) is considered as an Nx1 matrix.
        A 2-D array of shape (M, N) is considered as an MxN matrix.
    b (Variable): The right operand of the matrix multiplication.
        Its array is treated as a matrix in the same way as ``a`` 's array.
    ...
```

File: chainer/functions/matmul.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        transa (bool): If true, transpose a.
+        transb (bool): If true, transpose b.
+
+    Returns:
+        ~chainer.Variable: Matrix multiplication as a 2-D array
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> **The result of the** matrix multiplication**.**

File: chainer/functions/matmul.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+                          transb=not self.transb, transout=self.transa, out=ga)
+        _batch_matmul_gpu(a, gy[0],
+                          transa=not self.transa, transout=self.transb, out=gb)
+        ga = ga.reshape(a.shape)
+        gb = gb.reshape(b.shape)
+        return ga, gb
+
+
+def batch_matmul(a, b, transa=False, transb=False):
+    """Compute matrix multiplication of two arrays in a batch manner.
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> Compute**s** **the** matrix multiplication**s** of two **sets of** arrays in a batch manner.

File: chainer/functions/matmul.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+    Args:
+        a, b: Variables of 3-D or 2-D arrays.
+            A 3-D array of shape (B, N, M) is considered as B NxM matrices.
+            A 2-D array of shape (B, N,) is considered as B Nx1 matrices.
```
Comment (beam2d) [inline chainer/functions/matmul.py]: Split the entry to two entries for `a` and `b` as same as matmul.

File: chainer/functions/matmul.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        transa (bool): If true, transpose each matrix in a.
+        transb (bool): If true, transpose each matrix in b.
+
+    Returns:
+        ~chainer.Variable: Batch of matrix multiplications as a 3-D array.
```
Comment (beam2d) [inline chainer/functions/matmul.py]: -> **The result** of **the batch** matrix multiplications.

File: tests/functions_tests/test_matmul.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        y = self.op(x1, x2)
+        if isinstance(y.data, cuda.GPUArray):
+            self.assertTrue(hasattr(y.data.gpudata, 'device'))
+        gradient_check.assert_allclose(self.forward_answer, y.data)
+
+    def forward_cpu(self):
+        self.check_forward(self.x1, self.x2)
+
+    def test_matmul_forward_cpu(self):
+        self.forward_cpu()
```
Comment (beam2d) [inline tests/functions_tests/test_matmul.py]: Would you merge the test_matmul_forward_cpu and forward_cpu?

File: tests/functions_tests/test_matmul.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+    @attr.gpu
+    def forward_gpu(self):
+        self.check_forward(cuda.to_gpu(self.x1), cuda.to_gpu(self.x2))
+
+    @attr.gpu
+    def test_matmul_forward_gpu(self):
+        self.forward_gpu()
```
Comment (beam2d) [inline tests/functions_tests/test_matmul.py]: ditto

File: tests/functions_tests/test_matmul.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        gx1, gx2 = gradient_check.numerical_grad(
+            f, (x1.data, x2.data), (y.grad,))
+        gradient_check.assert_allclose(gx1, x1.grad, atol=atol)
+        gradient_check.assert_allclose(gx2, x2.grad, atol=atol)
+
+    def backward_cpu(self, atol=1e-2):
+        self.check_backward(self.x1, self.x2, self.gy, atol)
+
+    def test_matmul_backward_cpu(self):
+        self.backward_cpu()
```
Comment (beam2d) [inline tests/functions_tests/test_matmul.py]: ditto

File: tests/functions_tests/test_matmul.py [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+    @attr.gpu
+    def backward_gpu(self, atol=1e-2):
+        self.check_backward(
+            cuda.to_gpu(self.x1), cuda.to_gpu(self.x2),
+            cuda.to_gpu(self.gy), atol)
+
+    @attr.gpu
+    def test_matmul_backward_gpu(self):
+        self.backward_gpu()
```
Comment (beam2d) [inline tests/functions_tests/test_matmul.py]: ditto

**Comment by beam2d** [comment]
I added some comments, so please check them.

**Comment by muupan** [comment]
Thank you for your comments. I fixed all the points you mentioned.

File: chainer/functions/matmul.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
+import ctypes
+import numpy
```
Comment (beam2d) [inline chainer/functions/matmul.py]: Add a blank line between imports for standard libraries and third-party libraries.

**Comment by beam2d** [comment]
LGTM except above one comment!

**Comment by muupan** [comment]
Now it's fixed.

**Comment by beam2d** [comment]
[LGTM](http://lgtm.in/i/WXZaAe1Yp)

**Comment by beam2d** [comment]
Thank you!!!

**Comment by okuta** [comment]
fix #8

**Pull request merged**